### PR TITLE
storage-engine:  Add S3 as a storage location with rocksdb-cloud

### LINKED ISSUE #8367: Demo TiKV performance on S3
## Development Task

This task a subtask of https://github.com/tikv/tikv/issues/6506

Update tikv to be able to use S3 as storage, using the rocksdb-cloud integration that is done in https://github.com/tikv/rust-rocksdb/pull/517. Task break down

- [ ] Update to latest of tikv/rust-rocksdb, enable the `cloud` feature.
- [ ] Add a `S3Config` field to the `DbConfig` struct to pass S3 related options. The `S3Config` struct should also have a boolean `enabled` field, similar to what `TitanConfig` does. https://github.com/tikv/tikv/blob/master/src/config.rs#L874
- [ ] if `s3.enabled = true`, construct `AwsEnv` and pass to rocksdb similar to how we do to pass the `EncryptedEnv`to rocksdb: https://github.com/tikv/tikv/blob/master/cmd/src/server.rs#L333
- [ ] (This is probably the most difficult step. Suggest to do as the last step.) Update tikv build scripts to build with aws sdk with s3 support.

### PR DESCRIPTION
**Pull request by de-sh**
### What problem does this PR solve?

Issue Number: closes #8367 

Problem Summary: Update tikv to be able to use S3 as storage, using the rocksdb-cloud integration that is done in tikv/rust-rocksdb#517. Task break down

### Work plan

- [x] Update to using latest tikv/rust-rocksdb with cloud integrations.
- [x] Add `s3: S3Config` as field to the `DbConfig` struct to pass AWS S3 related options. Currently the `S3Config` struct is defined [here](https://github.com/tikv/tikv/pull/8383/files#diff-6d8a5e2104954a026de2bfe2b936aea4R813).
- [x] If `s3.enabled = true`, construct `AwsEnv` and pass to RocksEngine.
- [ ] Update build scripts to build with AWS-CPP-SDK for s3 support.

### Related changes

- PR to update `tik/rust-rocksdb`: tikv/rust-rocksdb#517
- PR to create `6.4.cloud` branch of `tikv/rocksdb` including only cloud integrations: tikv/rocksdb#182
- PR to update `6.4.tikv` branch of `tikv/rocksdb` to compile with cloud features: tikv/rocksdb#181

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -16,6 +16,7 @@ default = ["protobuf-codec"]
 tcmalloc = ["tikv_alloc/tcmalloc"]
 jemalloc = ["tikv_alloc/jemalloc", "engine_rocks/jemalloc"]
 mimalloc = ["tikv_alloc/mimalloc"]
+cloud = ["engine_rocks/cloud"]
```
Comment (yiwu-arbug) [inline Cargo.toml:19]: no need to gate with a feature.

File: cmd/src/server.rs [hunk 2]
```diff
@@ -361,7 +361,12 @@ impl TiKVServer {
     }
 
     fn init_engines(&mut self) {
-        let env = get_env(self.encryption_key_manager.clone(), None /*base_env*/).unwrap();
+        #[allow(unused_mut)]
+        let mut base_env = None;
+        if self.config.rocksdb.s3.enabled {
```
Comment (yiwu-arbug) [inline cmd/src/server.rs:366]: can we merge the logic into `get_env`?

Comment (de-sh) [reply cmd/src/server.rs:366]: Ok, right now get_env is defined in engine_rocks/encryption.rs

Comment (de-sh) [reply cmd/src/server.rs]: > can we merge the logic into `get_env`?

Can you check if the new usage is right...

**Comment by de-sh** [comment]
There's an error that pops up in `tikv.log` with respect to the use of cloud features when setup with tiup utility.
```
[FATAL] 
[lib.rs:483] 
["called `Result::unwrap()` on an `Err` value: Rocks(\"Not implemented: RocksDB Cloud not compiled with AWS support\")"] [backtrace="stack backtrace:
   0: tikv_util::set_panic_hook::{{closure}}
             at components/tikv_util/src/lib.rs:482
   1: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:524
   2: rust_begin_unwind
             at src/libstd/panicking.rs:431
   3: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   4: core::option::expect_none_failed
             at src/libcore/option.rs:1269
   5: core::result::Result<T,E>::unwrap
             at /home/devdutt/.rustup/toolchains/nightly-2020-07-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/result.rs:1005
   6: cmd::server::TiKVServer::init_engines
             at cmd/src/server.rs:374
   7: cmd::server::run_tikv
             at cmd/src/server.rs:96
   8: tikv_server::main
             at cmd/src/bin/tikv-server.rs:165
   9: std::rt::lang_start::{{closure}}
             at /home/devdutt/.rustup/toolchains/nightly-2020-07-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67
  10: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
      std::panicking::try::do_call
             at src/libstd/panicking.rs:342
      std::panicking::try
             at src/libstd/panicking.rs:319
      std::panic::catch_unwind
             at src/libstd/panic.rs:394
      std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  11: std::rt::lang_start
             at /home/devdutt/.rustup/toolchains/nightly-2020-07-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67
  12: main\n  13: __libc_start_main
  14: _start
"]
[location=cmd/src/server.rs:379]
[thread_name=main]
```
To reproduce it, use the PR and include the following configs in the `tikv.toml` config file:
```
[rocksdb.s3]
enabled = true
src-cloud-bucket = "{bucket_name}"
src-cloud-object = "{object_name}"
src-cloud-region = "{region_name}"
dest-cloud-bucket = "{bucket_name}"
dest-cloud-object = "{object_name}"
dest-cloud-region = "{region_name}"
```

and copying the executable that is generated from the above PR to replace the tikv-server being used by the default tiup cluster.

**Comment by gregwebs** [comment]
closing in favor of #8786

**Pull request closed without merging**